spinand.c

```diff
@@ -0,0 +1,279 @@
+                          unsigned int len, unsigned int chunksize)
+{
+	/* Read data over multiple pages */
+	int ret;
+
+	/* Calculate acture column address len */
+	unsigned int column_addr_len = ilog2(flash->chip->page_size);
+	unsigned int column_addr_mask = ~(-1 << column_addr_len);
+	printf("%d %d %d\n",flash->chip->page_size,column_addr_len,column_addr_mask);
+
+	while (len) {
+		unsigned int data_to_read = min(chunksize, len);
+		ret = spi_nand_inpage_read(flash, start >> column_addr_len, start & column_addr_mask, buf, len, data_to_read);
+		if (ret)
+			return ret;
+		len -= data_to_read;
+		start += data_to_read;
+		buf += data_to_read;
+		printf("chunked %d\n",len);
+	}
+
+	return 0;
+}
+
+/***************************************
+ * PROGRAM CYCLE
+ ***************************************/
+
+
+/***************************************
+ * PROBE
+ ***************************************/
+
+static char *strncpy_tospace(char *dest, const char *src, unsigned int len)
+{
+	int i;
+
+	for (i = 0; i < len; i++) {
+		dest[i] = src[i];
+		if (dest[i] == ' ') {
+			dest[i] = '\0';
+			break;
+		}
+	}
+	return dest;
+}
+
+static void probe_spi_nand_toshiba(struct flashctx *flash, const uint8_t *parameters)
+{
+	char *vendor = (char *)malloc(12);
+	char *name = (char *)malloc(20);
+
+	flash->chip->vendor = strncpy_tospace(vendor, (const char *)(parameters + 32), sizeof(vendor));
+	flash->chip->name = strncpy_tospace(name, (const char *)(parameters + 44), sizeof(name));
+	flash->chip->page_size = *(uint32_t *)(parameters + 80);
+	flash->chip->data_per_spare = flash->chip->page_size / *(uint16_t *)(parameters + 84);
+	flash->chip->partial_page_size = *(uint32_t *)(parameters + 86);
+	flash->chip->block_size = flash->chip->page_size * *(uint32_t *)(parameters + 92);
+	flash->chip->unit_size = flash->chip->block_size * *(uint32_t *)(parameters + 96);
+	flash->chip->total_size = flash->chip->unit_size * parameters[100] / 1024;
+	flash->chip->total_size = 64;
+}
+
+#define HAS_MAGIC(data, magic) (strncmp((const char *)magic, (const char *)data, sizeof(magic)) == 0)
+
+int probe_spi_nand(struct flashctx *flash)
+{
+	/* Read the Parameter Page */
+	uint8_t parameters[SPINAND_MAX_PARAMETER_PAGE_SIZE];
+	int i;
+
+	uint8_t feature_b0 = spi_nand_get_feature(flash, 0xB0);
+	spi_nand_set_feature(flash, 0xB0, feature_b0 | SPINAND_FEATURE_B0_IDR_E);
+	spi_nand_inpage_read(flash, 0x01, 0x00, parameters, SPINAND_MAX_PARAMETER_PAGE_SIZE, 0);
+	spi_nand_set_feature(flash, 0xB0, feature_b0);
+
+	msg_cdbg("SPI NAND returned");
+	for (i = 0; i < sizeof(parameters); i++)
+		msg_cdbg(" %02x", parameters[i]);
+	msg_cdbg("\n");
+
+	if (HAS_MAGIC(parameters, spi_nand_magic_toshiba)) {
+		msg_cdbg("%s: Found Toshiba\n", __func__);
+		probe_spi_nand_toshiba(flash, parameters);
+		return 1;
+	}
+
+	if (HAS_MAGIC(parameters, spi_nand_magic_micron)) {
+		msg_cdbg("%s: Found Micron\n", __func__);
+		probe_spi_nand_toshiba(flash, parameters);
+		return 1;
+	}
```

Comment:
This is not implemented?

Reply:
Well, it is, but it falls back to toshiba. Does Toshiba `NAND` magic implement ONFI?

Reply:
The only chip I have is Toshiba TC58CVG1S3HRAIG. Micron "support" is only here for PR demonstration.